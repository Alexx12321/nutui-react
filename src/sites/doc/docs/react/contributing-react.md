## 贡献指南

您可根据《[NutUI 邀您共建，Contributor 您好](https://jelly.jd.com/article/6320528b92d94a0068685525)》中的共建方向为 NutUI 社区做出贡献。

&nbsp;

本篇指南会指导你如何为 NutUI 贡献一份自己的力量，请在您要提 issue 或者 pull request 之前花几分钟来阅读一遍这篇指南，保证贡献是符合规范并且能帮助到社区。

## 行为准则

为保证良好的网络环境，营造舒适的开发氛围，希望所有的贡献者都能遵守这份[行为准则](https://www.contributor-covenant.org/zh-cn/version/2/1/code_of_conduct/)，也请花时间阅读一遍全文明确哪些行为是禁止的，哪些行为是可行的。

## 分支管理

目前 github 上长期维护分支 main。如果您要修一个 bug 或新增功能，请发 pull request 到 main 分支。

&nbsp;

官网会同步更新 [NutUI-React 的发布周期](https://github.com/jdf2e/nutui-react/releases)，同时每天都会有开发者 review github 中提交的代码。发版周期是每 1 周/次，通常会提前合并新的代码或功能，发布正式包；期间如遇紧急需求，可基于指定分支发布 beta 版本。

## Issue 报告指南

如果提交的是 Bug 报告，请务必遵守 [Bug report](https://github.com/jdf2e/nutui/blob/next/.github/ISSUE_TEMPLATE/bug_report.md) 模板。

## Bugs

我们使用 [issue-helper](https://nutui.jd.com/nutui-issue-helper/?repo=jdf2e/nutui-react) 来做 bug 追踪。希望大家通过我们提供的 issue 小助手来提交反馈，并尽可能提供一切开发信息，因为信息越全面，开发响应速度越快，您发现的 bug 会被快速解决。

&nbsp;

提交 bug 之前，请先确保搜索过已有的 [issue](https://github.com/jdf2e/nutui-react/issues) 和阅读了我们的[常见问题](https://nutui.jd.com/#/zh-CN/notice)。

## 功能新增

如果您对现有的组件功能或者 API 有更好的想法，我们同样推荐你使用我们提供的 [issue-helper](https://nutui.jd.com/nutui-issue-helper/?repo=jdf2e/nutui-react) 提交一个添加新功能的 issue。

&nbsp;

如果你希望协助开发新功能新组件，可扫描[二维码](https://img13.360buyimg.com/imagetools/s300x300_jfs/t1/187308/9/29678/349561/634fa983E93012590/a641089fdfed5911.png)加入我们的社区共建群。

## 参与贡献

如果你对 github 的 PR 流程还不熟悉，可以阅读下面这 2 篇文章来学习：

&nbsp;

[NutUI 贡献指南](https://github.com/jdf2e/nutui-react/discussions/348)

[github 仓库代码同步](https://github.com/jdf2e/nutui/discussions/687)

## Pull Request 规范

1. 每个 PR 只解决单个问题或者新增单个功能

2. 当新增组件或修改原有组件时，同步更改对应的单元测试和文档，保证代码的稳定

3. commit message 请遵循 [Angular Style Commit Message Conventions](https://gist.github.com/stephenparish/9941e89d80e2bc58a153) 规范

4. PR 信息中最好包括修改了哪个组件，解决了哪个问题，并关联 github 上对应的 issue 链接

## Pull Request 流程

1、fork 主仓库，并同步主仓库的最新代码

2、将项目下载到本地，在项目根目录下运行 npm install，并在本地能正常访问

```
git clone https://github.com/{github username}/nutui-react.git
npm install
npm run dev
```

3、创建一个新的分支或在对应分支上进行开发

```
git checkout -b username/xxxx
git checkout {现有分支名称}
```

4、如开发涉及新组件或组件的新特性，需要完善组件的中英文文档、组件的单元测试，同时要使用 [nutui-demo](https://github.com/jdf2e/nutui-demo) 对组件测试使用。

5、保证编译功能正常使用

```
npm run test
// build 使用 node 17 版本
// 构建 @nutui/nutui-react
npm run build
// 构建 @nutui/nutui-react-taro
npm run build:taro
```

6、开发完成后提交到仓库的对应分支上，PR 会在 Review 通过后被合并到主仓库，然后等待 NutUI-React 发布新版本（每周一次）。

### 知识分享

1. 项目使用案例

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;分享 NutUI 的[使用案例](https://nutui.jd.com/next/#/case)，我们会在官网进行推荐展示，准备项目的名称、项目入口（二维码\链接）、logo、示例图、投放平台（H5\小程序\APP等）、nutui 版本等信息，以邮箱的形式提交到 **nutui@jd.com**。

&nbsp;

2. 文章视频投稿

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;分享你的经验（教程、文章、视频等），可以给 **nutui@jd.com** 投稿。在审核通过后，收藏到[官网资源](https://nutui.jd.com/next/#/resource)中，将你的知识与思想传递给 NutUI 的小伙伴们。

&nbsp;

3. 共建分享征集

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;以 NutUI-React 为基础，设计或开发新的工具，开发提效的插件等。有兴趣欢迎加入我们的[社区开发群](https://img13.360buyimg.com/imagetools/s300x300_jfs/t1/187308/9/29678/349561/634fa983E93012590/a641089fdfed5911.png)交流讨论。

&nbsp;

### Credits

感谢以下所有给 NutUI 贡献过代码的开发者。
&nbsp;
<a href="https://github.com/jdf2e/nutui-react/graphs/contributors">
<img src="https://opencollective.com/nutui-react/contributors.svg?width=920&button=false" alt="contributors"/>
</a>
