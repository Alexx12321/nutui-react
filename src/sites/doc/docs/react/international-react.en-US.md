# Internationalization

NutUI-React support multiple languages. NutUI uses Chinese as the default language. Use as follows:

## Usage

### Switch languages

:::demo

```javascript
import { ConfigProvider } from '@nutui/nutui-react'
import en from '@nutui/nutui-react/dist/locales/en-US'

ReactDOM.render(
  <ConfigProvider locale={en}>
    <App />
  </ConfigProvider>,
  document.querySelector('#root')
)
```

:::

## Current supported languages:

| Language | Filename | Version |
| --- | --- | --- |
| English | en-US | `v1.0.0` |
| Chinese | zh-CN | `v1.0.0` |
| Uygur | zh-UG | `v1.3.15` |
| Traditional Chinese | zh-TW | `v1.0.0` |
| Indonesian | id-ID | `v1.3.2` |
| Thai | th-TH | PR |

> View all language configs [Here](https://github.com/jdf2e/nutui-react/tree/main/src/locales).

## Common Questions

### Missing language pack?

Welcome to commit PR If you need new language pack. Changes refer to [language pack](https://github.com/jdf2e/nutui-react/tree/main/src/locales).
