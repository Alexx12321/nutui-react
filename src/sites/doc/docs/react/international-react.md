# 国际化

NutUI-React 1.0 以上版本支持多语言。组件默认使用中文。

## 使用方法

### 多语言切换

:::demo

```javascript
// 多语言版本
import { ConfigProvider } from '@nutui/nutui-react'
import en from '@nutui/nutui-react/dist/locales/en-US'

ReactDOM.render(
  <ConfigProvider locale={en}>
    <App />
  </ConfigProvider>,
  document.querySelector('#root')
)
```

:::

## 目前支持的语言:

| 语言 | 文件名 | 版本 |
| --- | --- | --- |
| 英语 | en-US | `v1.0.0` |
| 简体中文 | zh-CN | `v1.0.0` |
| 维吾尔语 | zh-UG | `v1.3.15` |
| 繁體中文 | zh-TW | `v1.0.0` |
| 印度尼西亚语 | id-ID | `v1.3.2` |
| 泰语 | th-TH | 等待 PR |

> 在 [这里](https://github.com/jdf2e/nutui-react/tree/main/src/locales) 查看所有的语言包源文件。

## 常见问题

### 找不到所需的语言包？

如果上方列表中没有你需要的语言，欢迎给我们提 Pull Request 来增加新的语言包。改动内容可以参考 [语言包](https://github.com/jdf2e/nutui-react/tree/main/src/locales) 的 PR
