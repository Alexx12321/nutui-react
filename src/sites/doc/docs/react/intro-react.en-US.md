# NutUI-React

The NutUI-React component library is suitable for H5 applications of the React technology stack. It can be used out of the box to help developers quickly develop user interfaces, improve development efficiency, and improve development experience.

## Features

- 🚀 70+ high-quality components covering mainstream mobile scenarios
- 📖 Based on JD APP 10.0 visual specification
- 🍭 Support on-demand reference
- 📖 Detailed documentation and examples
- 💪 Supports TypeScript
- 💪 Support for server-side rendering (Next.js)
- 🍭 Support custom themes
- 🌍 Internationalization support
- 🍭 Unit test coverage exceeds 80% to ensure stability
- 📖 Sketch design resources are available

## Examples

<img src="https://storage.360buyimg.com/imgtools/cc08cb2760-d70f19c0-0a95-11ee-808d-b1254110adba.png" width="200" alt="NutUI-React" />

## Compatibility Environment

| Version | Supported Frameworks | Visual Specifications | Compatibility |
| --- | --- | --- | --- |
| @nutui/nutui-react <img src="https://img.shields.io/npm/v/@nutui/nutui-react" /> | React 17\18 | Jingdong [APP 10.0](#/resource) | Modern browsers start with Chrome >= 51、iOS >= 10.0 |

For some older versions of browsers, corresponding polyfills are needed. It is recommended to use @babel/preset-env to address browser compatibility issues. The package provided by NutUI-React has not undergone extensive syntax transformation, so it is necessary to include the NutUI-React package in the compilation phase.

For browsers that do not support CSS variables, it is recommended to use [postcss-var-replace](https://www.npmjs.com/package/postcss-var-replace) to resolve compatibility issues.

## All Contributors

Thanks to the following friends for their contributions to NutUI-React:

<a href="https://github.com/jdf2e/nutui-react/graphs/contributors">
  <img src="https://opencollective.com/nutui-react/contributors.svg?width=890&button=false" alt="contributors">
</a>

## LICENSE

[MIT](https://zh.wikipedia.org/wiki/MIT%E8%A8%B1%E5%8F%AF%E8%AD%89)
