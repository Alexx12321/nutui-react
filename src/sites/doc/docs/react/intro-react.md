# NutUI-React

NutUI-React 组件库，适用于 React 技术栈的 H5 应用，开箱即用，帮助研发快速开发用户界面，提升开发效率，改善开发体验。

## 特性

- 🚀 80+ 高质量组件，覆盖移动端主流场景
- 📖 基于京东APP 10.0 视觉规范
- 🍭 支持按需引用
- 📖 详尽的文档和示例
- 💪 支持 TypeScript
- 💪 支持服务端渲染（Next.js）
- 🍭 支持定制主题
- 🌍 国际化支持
- 🍭 单元测试覆盖率超过 80%，保障稳定性
- 📖 提供 Sketch 设计资源

## 扫码体验

<img src="https://storage.360buyimg.com/imgtools/cc08cb2760-d70f19c0-0a95-11ee-808d-b1254110adba.png" width="200" alt="NutUI-React" />

## 兼容情况

| 版本 | 支持的框架 | 视觉规范 | 兼容 |
| --- | --- | --- | --- |
| @nutui/nutui-react <img src="https://img.shields.io/npm/v/@nutui/nutui-react" /> | React 17\18 | 京东 [APP 10.0](#/resource) 规范 | 现代浏览器以 Chrome >= 51、iOS >= 10.0 |

对于一些低版本的浏览器，需要相应的 Ployfill，建议使用 @babel/preset-environment 来解决浏览器兼容问题。NutUI-React 提供的包并未进行过多的语法转换，所以需要将 NutUI-React 包含到编译阶段。

对于不支持 CSS 变量的浏览器，建议使用 [postcss-var-replace](https://www.npmjs.com/package/postcss-var-replace) 来解决兼容问题。

## 贡献者们

感谢以下小伙伴们为 NutUI-React 发展做出的贡献：

<a href="https://github.com/jdf2e/nutui-react/graphs/contributors">
  <img src="https://opencollective.com/nutui-react/contributors.svg?width=890&button=false" alt="contributors" />
</a>

## 开源协议

本项目基于 [MIT](https://zh.wikipedia.org/wiki/MIT%E8%A8%B1%E5%8F%AF%E8%AD%89) 协议
