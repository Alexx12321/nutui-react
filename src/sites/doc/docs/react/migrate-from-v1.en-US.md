# Upgrading from v1 to v2

This document will help you upgrade from NutUI React `1.x` to NutUI React `2.x` version.

## Upgrade Steps

1. H5 Install NutUI React version 2.x

```shell
npm install @nutui/nutui-react
```

2. Taro installs NutUI React version 2.x

```shell
npm install @nutui/nutui-react-taro
```

3. Handling incompatible updates
   There are some incompatible updates from NutUI React 1.x to NutUI React 2.x. You need to read the incompatible updates carefully and deal with them in turn.
   You can manually check your code against the list below to make changes, or we provide a codemod cli tool @nutui/nutui-react-codemod to help you upgrade to v2 quickly. Please commit your local code changes before running codemod cli.
4. Theme variables are renamed:
   For example, primary-color is renamed to color-primary; note that if you are using a custom theme, especially if you are using the ConfigProvider component, you are not using `nutuiBrandColor`, remember to rename it to `nutuiColorPrimary`.

## Compatibility Updates

1. Component style handling
   Added support for importing css files on demand, while retaining the ability to import scss files on demand. On-demand css import can be realized by babel-import-plugin plugin:
   The H5 configuration is as follows:

```json
// Webpack .babelrc or babel.config.js configuration
plugins: [
  [
    “import”.
    {
      libraryName: “@nutui/nutui-react”, libraryDirectory: “dist/esm”, {
      libraryDirectory: “dist/esm”, {
      style: 'css',
      camel2DashComponentName: false,
    }.
    “nutui-react”.
  ]
]
```

Taro is configured as follows:

```json
// Configured in Webpack .babelrc or babel.config.js
plugins: [
  [
    “import”.
    {
      libraryName: “@nutui/nutui-react-taro”, libraryDirectory: “dist/esm”, {
      libraryDirectory: “dist/esm”, {
      style: 'css',
      camel2DashComponentName: false,
    }.
    “nutui-react-taro”.
  ]
]
```

2. better type exporting and adding `JSDoc` annotations to types
3. Adjustment of component categorization
   In terms of component classification, we have reviewed the 1.x classification based on the information structure from the interaction dimension, together with the interaction design side, and subclassed and reclassified them, with the goal of more closely matching the distribution of interaction scenarios and making it easy to find components. The main distribution is:

- Basic components, remove the `Popup` component, and subdivide `Popup` into operation feedback-guidance tips section;
- Layout components, remain unchanged;
- Navigation components: move paging related components `Pagination` and `Indicator` to the display component (to consider the light operation of paging on mobile); move `Menu` menu to the subclass of data entry class-selector (to consider that `Menu` is mainly used as a filter);

Translated with DeepL.com (free version)
