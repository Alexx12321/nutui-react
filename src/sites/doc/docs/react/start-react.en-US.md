# Quick Start

## Intro

Through this article, you can master the installation and use of NutUI React. The operation is simple and easy to use, and the development is simple and fast.

## Install

#### 1. Install via NPM

```sh
#pnpm
pnpm add @nutui/nutui-react
#yarn
yarn add @nutui/nutui-react
# npm
npm install @nutui/nutui-react
```

#### 2. Install and use via CDN

> NutUI is available on public CDNs like **jsdelivr** and **unpkg**.
> It is not recommended to use the component library CDN in the production environment. If this method is required, it is recommended to download the specific version of the CDN file to the local project directory for use.

:::demo

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <!-- import style -->
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/@nutui/nutui-react/dist/style.css"
    />
    <!-- import React -->
    <script
      crossorigin
      src="https://unpkg.com/react@17/umd/react.production.min.js"
    ></script>
    <script
      crossorigin
      src="https://unpkg.com/react-dom@17/umd/react-dom.production.min.js"
    ></script>
    <!-- Import NutUI component library -->
    <script src="https://cdn.jsdelivr.net/npm/@nutui/nutui-react/dist/nutui.react.umd.js"></script>
  </head>
  <body>
    <div id="app"></div>
    <script>
      // Render a button component under the #app tag
      ReactDOM.render(
        React.createElement(
          nutui.react.Button,
          null,
          React.createElement('div', null, 'Primary button')
        ),
        document.querySelector('#app')
      )
    </script>
  </body>
</html>
```

:::

## Component usage

> NutUI React supports Tree Shaking by default, and supports on-demand import of component JS files without any plug-in configuration. However, css style files cannot be implemented in this way, so from the perspective of style, it can be divided into the following two usage methods:

#### Method 1, Conventional usage: Full import of styles

:::demo

```js
import * as React from 'react'
import * as ReactDOM from 'react-dom'
import '@nutui/nutui-react/dist/style.css'
import { Button } from '@nutui/nutui-react'

ReactDOM.render(
  <div className="App">
    <Button></Button>
  </div>,
  document.getElementById('app')
)
```

:::

#### Method 2, import styles on demand

<details>
<summary>vite</summary>

Install the `vite-plugin-imp` plugin and configure it.

```sh
#pnpm
pnpm add vite-plugin-imp -D
#yarn
yarn add vite-plugin-imp -D
# npm
npm install vite-plugin-imp -D
```

:::demo

```js
// vite.config.js
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'
import vitePluginImp from 'vite-plugin-imp'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    react(),
    vitePluginImp({
      libList: [
        {
          libName: '@nutui/nutui-react',
          style: (name) => {
            return `@nutui/nutui-react/dist/esm/${name}/style/css`
          },
          replaceOldImport: false,
          camel2DashComponentName: false,
        },
      ],
    }),
  ],
})
```

:::

</details><br/>

<details>
<summary>webpack</summary>

Install the `babel-plugin-import` plugin and configure it.

```sh
#pnpm
pnpm add babel-plugin-import -D
#yarn
yarn add babel-plugin-import -D
# npm
npm install babel-plugin-import -D
```

Babel configuration:

:::demo

```js
{
  //...
  plugins: [
    [
      'import',
      {
        libraryName: '@nutui/nutui-react',
        libraryDirectory: 'dist/esm',
        style: 'css',
        camel2DashComponentName: false,
      },
      'nutui-react',
    ],
  ]
}
```

:::

</details><br/>

## Precautions for use

- NutUI-React is built on [react@^18.0.0](https://www.npmjs.com/package/react)
- The component CSS unit uses **px**, if you need **rem** unit in your project, you can convert it with some tools, such as [px2rem-loader](https://www.npmjs.com/package/px2rem-loader) of [webpack](https://www.webpackjs.com/), [postcss] of [postcss](https://github.com/postcss/postcss) -plugin-px2rem](https://www.npmjs.com/package/postcss-plugin-px2rem) plugin etc.

## Examples

Source code for all the examples can be found in the [packages/templates](https://github.com/jdf2e/nutui-react/tree/next/packages/nutui-templates) directory.

<content-examples/>
