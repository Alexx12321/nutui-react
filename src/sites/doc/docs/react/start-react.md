# 快速上手

## 介绍

通过本文你可以掌握 NutUI React 的安装和使用方法，操作简单易上手，开发简洁快速。

## 安装

#### 1. 通过 NPM 安装

```sh
# pnpm
pnpm add @nutui/nutui-react
# yarn
yarn add @nutui/nutui-react
# npm
npm install @nutui/nutui-react
```

#### 2. 通过 CDN 安装及使用

> 可以在 **jsdelivr** 和 **unpkg** 等公共 CDN 上获取到 NutUI。
> 不推荐在生产环境使用组件库 CDN，如果需要这种使用方式，建议将特定版本的 CDN 文件下载至本地项目目录中使用。

:::demo

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <!-- 引入样式 -->
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/@nutui/nutui-react/dist/style.css"
    />
    <!-- 引入React -->
    <script
      crossorigin
      src="https://unpkg.com/react@17/umd/react.production.min.js"
    ></script>
    <script
      crossorigin
      src="https://unpkg.com/react-dom@17/umd/react-dom.production.min.js"
    ></script>
    <!-- 引入NutUI组件库 -->
    <script src="https://cdn.jsdelivr.net/npm/@nutui/nutui-react/dist/nutui.react.umd.js"></script>
  </head>
  <body>
    <div id="app"></div>
    <script>
      // 在 #app 标签下渲染一个按钮组件
      ReactDOM.render(
        React.createElement(
          nutui.react.Button,
          null,
          React.createElement('div', null, '主要按钮')
        ),
        document.querySelector('#app')
      )
    </script>
  </body>
</html>
```

:::

## 组件使用

> NutUI React 默认支持 Tree Shaking，在无任何插件配置的情况下支持组件 JS 文件的按需引入。但 css 样式文件无法通过这种方式实现，因此从样式的角度可以分为以下两种使用方法：

#### 方法一、常规用法：全量引入样式

:::demo

```js
import * as React from 'react'
import * as ReactDOM from 'react-dom'
import '@nutui/nutui-react/dist/style.css'
import { Button } from '@nutui/nutui-react'

ReactDOM.render(
  <div className="App">
    <Button></Button>
  </div>,
  document.getElementById('app')
)
```

:::

#### 方法二、按需引入样式

<details>
<summary>vite</summary>

安装 `vite-plugin-imp` 插件并配置。

```sh
# pnpm
pnpm add vite-plugin-imp -D
# yarn
yarn add vite-plugin-imp -D
# npm
npm install vite-plugin-imp -D
```

:::demo

```js
// vite.config.js
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'
import vitePluginImp from 'vite-plugin-imp'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    react(),
    vitePluginImp({
      libList: [
        {
          libName: '@nutui/nutui-react',
          style: (name) => {
            return `@nutui/nutui-react/dist/esm/${name}/style/css`
          },
          replaceOldImport: false,
          camel2DashComponentName: false,
        },
      ],
    }),
  ],
})
```

:::

</details><br />

<details>
<summary>webpack</summary>

安装 `babel-plugin-import` 插件并配置。

```sh
# pnpm
pnpm add babel-plugin-import -D
# yarn
yarn add babel-plugin-import -D
# npm
npm install babel-plugin-import -D
```

babel 配置：

:::demo

```js
{
  // ...
  plugins: [
    [
      'import',
      {
        libraryName: '@nutui/nutui-react',
        libraryDirectory: 'dist/esm',
        style: 'css',
        camel2DashComponentName: false,
      },
      'nutui-react',
    ],
  ]
}
```

:::

</details><br/>

## 使用注意事项

- NutUI-React 基于 [react@^18.0.0](https://www.npmjs.com/package/react) 构建
- 组件 CSS 单位使用的是 **px**，如果你的项目中需要 **rem** 单位，可借助一些工具进行转换，比如 [webpack](https://www.webpackjs.com/) 的 [px2rem-loader](https://www.npmjs.com/package/px2rem-loader)、[postcss](https://github.com/postcss/postcss) 的 [postcss-plugin-px2rem](https://www.npmjs.com/package/postcss-plugin-px2rem) 插件等

## 示例

示例代码可以再 [packages/templates](https://github.com/jdf2e/nutui-react/tree/next/packages/nutui-templates) 下查看

<content-examples/>
