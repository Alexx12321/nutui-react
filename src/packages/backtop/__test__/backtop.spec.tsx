// import * as renderer from 'react-test-renderer'
import * as React from 'react'
import '@testing-library/jest-dom'
import { Top } from '@nutui/icons-react'
import { act, fireEvent, render, waitFor } from '@testing-library/react'
import BackTop from '@/packages/backtop'

test('backtop props test', () => {
  const handleClick = vi.fn()
  const { container } = render(
    <div id="target" style={{ height: '100vh' }}>
      {new Array(24).fill(0).map((_, index) => {
        return <div key={index}>我是测试数据{index}</div>
      })}
      <BackTop
        target="target"
        className="backtop-button"
        onClick={handleClick}
      />
    </div>
  )
  const chooseTagEle = container.querySelectorAll('.backtop-button')[0]
  fireEvent.click(chooseTagEle)
  expect(handleClick).toBeCalled
})

test('backtop custom test', () => {
  const handleClick = vi.fn()
  const { container } = render(
    <BackTop
      className="custom-class"
      target="target"
      threshold={100}
      style={{
        bottom: '110px',
        right: '10px',
      }}
      onClick={handleClick}
    >
      <Top />
      <div style={{ fontSize: '12px' }}>顶部</div>
    </BackTop>
  )
  expect(container.querySelector('.nut-backtop')).toHaveAttribute(
    'style',
    'z-index: 900; bottom: 110px; right: 10px;'
  )
  fireEvent.click(container)
  expect(handleClick).toBeCalled
  expect(container).toMatchSnapshot()
})

test('scroll', async () => {
  const { container } = render(
    <div id="target" style={{ height: '100px' }} className="backtop-wrapper">
      {new Array(24).fill(0).map((_, index) => {
        return (
          <div key={index} style={{ height: 30 }}>
            我是测试数据{index}
          </div>
        )
      })}
      <BackTop target="target" className="backtop-button" />
    </div>
  )
  const track = container.querySelector('.backtop-wrapper')
  const element18 = container.querySelectorAll(
    '.nut-hoverbutton-item-container'
  )[0]
  const element19 = container.querySelectorAll('.nut-hoverbutton-container')[0]
  if (track) {
    track.scrollTo = vi.fn()
    track.scrollTop = 200
    act(() => {
      track.dispatchEvent(new Event('scroll'))
    })
    await waitFor(() => {
      expect(element19).toHaveClass('nut-backtop-show')
    })
    fireEvent.click(element18 as Element)
  }
})
