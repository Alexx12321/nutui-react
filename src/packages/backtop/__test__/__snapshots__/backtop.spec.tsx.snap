// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`backtop custom test 1`] = `
<div>
  <div
    class="nut-hoverbutton-container nut-backtop custom-class"
    style="z-index: 900; bottom: 110px; right: 10px;"
  >
    <div
      class="nut-hoverbutton"
    >
      <div
        class="nut-hoverbutton-item-container"
      >
        <svg
          aria-labelledby="Top"
          class="nut-icon nut-icon-Top "
          role="presentation"
          viewBox="0 0 44 44"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M43.042 18.666a.6.6 0 0 1 .062.846l-1.506 1.743a.6.6 0 0 1-.846.062L23.753 6.64V42.9a.6.6 0 0 1-.6.599H20.85a.6.6 0 0 1-.6-.6V6.64L3.25 21.318a.6.6 0 0 1-.846-.062L.898 19.512a.6.6 0 0 1 .062-.846L22.001.5z"
            fill="currentColor"
          />
        </svg>
        <div
          style="font-size: 12px;"
        >
          顶部
        </div>
      </div>
    </div>
    <div
      class="nut-safe-area nut-safe-area-position-bottom"
    />
  </div>
</div>
`;
